### ISSUE TITLE
WhitespaceAfter: new token SINGLE_LINE_COMMENT to catch missing whitespace checks around // comments

### ISSUE THREAD

**Issue by VeniceWish** [checkstyle/checkstyle#17723]
I have read check documentation: 
https://checkstyle.sourceforge.io/checks/whitespace/whitespaceafter.html
https://checkstyle.sourceforge.io/checks/whitespace/whitespacearound.html
https://checkstyle.sourceforge.io/checks/misc/trailingcomment.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

From: https://google.github.io/styleguide/javaguide.html?utm_source=chatgpt.com#s4.6.2-horizontal-whitespace
> Between any content and a double slash (//) which begins a comment. Multiple spaces are allowed.
Between a double slash (//) which begins a comment and the comment's text. Multiple spaces are allowed.

Code: 
```java
class Test {
  void testSlash() {
    int num1; // Good Slash
    int num2;// Bad Slash, Expected 1 Violation
    int num3; //Bad Slash, Expected 1 Violation
    int num4;//Bad Slash, Expected 2 Violation
  }//Bad Slash, Expected 2 Violation
}


```

Cli: 
```cmd
$ java -jar checkstyle-10.26.1-all.jar -c google_checks.xml Test.java 
Starting audit...
[WARN] Test.java:4:13: ';' is not followed by whitespace. [WhitespaceAfter]
[WARN] Test.java:6:13: ';' is not followed by whitespace. [WhitespaceAfter]
[WARN] Test.java:7:3: WhitespaceAround: '}' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
Audit done.
```

Formatter:
```cmd
$ java -jar google-java-format-1.28.0-all-deps.jar Test.java > FormattedCode.java
$ diff -Naru Test.java FormattedCode.java
--- Test.java   2025-09-01 22:31:51.425716700 +0800
+++ FormattedCode.java  2025-09-01 23:04:18.575558200 +0800
@@ -1,9 +1,8 @@
 class Test {
   void testSlash() {
     int num1; // Good Slash
-    int num2;// Bad Slash, Expected 1 Violation
-    int num3; //Bad Slash, Expected 1 Violation
-    int num4;//Bad Slash, Expected 2 Violation
-  }//Bad Slash, Expected 2 Violation
-
+    int num2; // Bad Slash, Expected 1 Violation
+    int num3; // Bad Slash, Expected 1 Violation
+    int num4; // Bad Slash, Expected 2 Violation
+  } // Bad Slash, Expected 2 Violation
 }
```

---

Problem: currently only `;` or `}` tokens trigger diagnostics, but there is no direct check on spacing around `//`, which leads to false negatives.

 As a workaround, we tried using TrailingComment with the following config:
```xml
<module name="TrailingComment">
  <property name="format" value="null"/>
  <property name="legalComment" value="^ .*$"/>
</module>

```
This produces output:
```cmd
Starting audit...
[WARN] Test.java:4:13: ';' is not followed by whitespace. [WhitespaceAfter]
[WARN] Test.java:5:15: Don't use trailing comments. [TrailingComment]
[WARN] Test.java:6:13: ';' is not followed by whitespace. [WhitespaceAfter]
[WARN] Test.java:6:14: Don't use trailing comments. [TrailingComment]
[WARN] Test.java:7:3: WhitespaceAround: '}' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] Test.java:7:4: Don't use trailing comments. [TrailingComment]
Audit done.
```

However, this kind of indirect workaround is not very desirable, and it also makes the configuration harder to understand and maintain. Moreover, it forces users to set <property name="format" value="null"/>, which prevents them from using TrailingComment for its real purpose (e.g., to restrict or forbid trailing comments).

Perhaps a better solution would be to add SLASH (line comment delimiter) as a token available to `WhitespaceAfter` and `WhitespaceAround` (ideally a new `WhitespaceBefore` to align the granularity of responsibilities with `WhitespaceAfter`), so that we can require a space before `// `when trailing code and a space after `//` before the comment text. This approach is cleaner, does not block `TrailingComment` from its intended use, and would reduce duplicate warnings and false negatives in the future.

---

**What is new:**
New accessible token: `TokenTypes.SINGLE_LINE_COMMENT`
It is not a default token of Check.

**Comment by romani** [comment]
@VeniceWish , please use `java -jar checkstyle-10.26.1-all.jar -c /google_checks.xml Test.java`

Attention to `/` before config. It will use config from jar, so we can be sure that you use latest Google style.

We can have multiple Check instances, we already have 
https://github.com/checkstyle/checkstyle/blob/f486dcacb9b6096bd28ce9b6b3d215f922f315e7/src/main/resources/google_checks.xml#L439-L443

We just need to add `id` properties to Check to let them be different. https://checkstyle.org/config.html#Id 
 Message also needs to be overriden to guide user on what is expected

**Comment by VeniceWish** [comment]
Thanks again for the explanation about using multiple Check instances with different messages — that does solve the issue of enforcing spacing after `//`.

I may not have explained clearly enough about the whitespace before `//` in trailing comments.
The current Google style configuration implicitly enforces spacing before `//` comments by checking for whitespace after symbols like `{`, `}` or `;`. This works in many cases, but it's an indirect workaround and might miss violations in other situations. Here's a minimal example to illustrate:
Code:
```java
class Test {
  void testSlash() {// Expected 1 violation and got 1
    int num1;// Expected 1 violation and got 1
    int num2 =
        num1// Expected 1 violation but got 0
            + 2;
  }// Expected 1 violation and got 1
}

```
Cli:
```
$ java -jar checkstyle-10.26.1-all.jar -c /google_checks.xml Test.java
Starting audit...
[WARN] Test.java:2:20: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] Test.java:3:13: ';' is not followed by whitespace. [WhitespaceAfter]
[WARN] Test.java:7:3: WhitespaceAround: '}' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
Audit done.
```

Formatter:
```
$ java -jar google-java-format-1.28.0-all-deps.jar Test.java > FormattedCode.java
$ diff -Naru Test.java FormattedCode.java
--- Test.java   2025-09-25 12:09:39.627866200 +0800
+++ FormattedCode.java  2025-09-25 12:09:48.846663300 +0800
@@ -1,8 +1,8 @@
 class Test {
-  void testSlash() {// Expected 1 violation and got 1
-    int num1;// Expected 1 violation and got 1
+  void testSlash() { // Expected 1 violation and got 1
+    int num1; // Expected 1 violation and got 1
     int num2 =
-        num1// Expected 1 violation but got 0
+        num1 // Expected 1 violation but got 0
             + 2;
-  }// Expected 1 violation and got 1
+  } // Expected 1 violation and got 1
 }
```

**Comment by romani** [comment]
We don't have WhitespaceBefore check
https://checkstyle.org/checks/whitespace/index.html

So it might  be good to have such Check with https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#SINGLE_LINE_COMMENT

**Comment by romani** [comment]
So issue is valid.
solution is not clear. 

We can at least improve  by usage of
https://checkstyle.org/checks/misc/trailingcomment.html#TrailingComment
Or 
https://checkstyle.org/checks/misc/todocomment.html#TodoComment

And space before, could be separate issue

**Comment by Atharv3221** [comment]
I am working on this

**Comment by Atharv3221** [comment]
@romani I tried to fix this with adding SINGLE_LINE_COMMENT token to WhitespaceAround its working well but,
for case `//// comment`
it sees as 
```
    |--SINGLE_LINE_COMMENT -> // [4:0]
    |   `--COMMENT_CONTENT -> // comment\n [4:2]
```
so getting violation for this too

**Comment by romani** [comment]
>We don't have WhitespaceBefore check

I think we just need new Check like this.
We can create new issue to make it and after implementation is done,  continue this issue to activate it in google_check.xml and make google style to cover this flase-negative.

**Comment by M-SaaD-H** [comment]
Hi @surajgojanur 
Are you still working on this?

**Comment by romani** [comment]
Fix is merged
